src/dune_pkg/archive_driver.ml

```diff
@@ -0,0 +1,75 @@
+open Stdune
+module Process = Dune_engine.Process
+open Fiber.O
+
+type t =
+  { bin : Path.t Lazy.t
+  ; suffixes : string list
+    (* Identify a file by its suffix rather than by its extension, since
+       ".tar.gz" files only have an extension of ".gz". *)
+  ; build_args : archive:Path.t -> target_in_temp:Path.t -> string list
+  }
+
+let make ~bin_name ~suffixes ~build_args =
+  let bin =
+    lazy
+      (match Bin.which ~path:(Env_path.path Env.initial) bin_name with
+       | Some x -> x
+       | None -> Dune_engine.Utils.program_not_found bin_name ~loc:None)
+  in
+  { bin; suffixes; build_args }
+;;
+
+let tar =
+  make
+    ~bin_name:"tar"
+    ~suffixes:[ ".tar"; ".tar.gz"; ".tgz"; ".tar.bz2"; ".tbz" ]
+    ~build_args:(fun ~archive ~target_in_temp ->
+      [ "xf"; Path.to_string archive; "-C"; Path.to_string target_in_temp ])
+;;
+
+let zip =
+  make ~bin_name:"unzip" ~suffixes:[ ".zip" ] ~build_args:(fun ~archive ~target_in_temp ->
+    [ Path.to_string archive; "-d"; Path.to_string target_in_temp ])
+;;
```

Comment:
Generally Windows doesn't have the `unzip` binary unlike most Linux and macos distributions. It can however be installed. On the other hand, I believe `tar.exe` can handle `.zip` files.

https://ss64.com/nt/tar.html

Similarly most modern `tar` implementations on other platforms (even freeBSD) support this. I don't even have `unzip` by default on NixOS, so perhaps we should consider letting `tar` do the work here instead.

Reply:
> Similarly most modern tar implementations on other platforms (even freeBSD) support this.

Experimentally, GNU tar doesn't support zip files. BSD tar (installable on some linux distros in the "libarchive" package) does, as does the tar that comes with macos (which is based on BSD tar according to its manpage). Most linux users will have GNU tar, so we can't rely on tar being able to extract zip files in all cases.

However you make a good point about windows and the BSDs. In these cases (and possibly on macos too) it would make sense to fall back to attempting to use tar for zip files if `unzip` is not in PATH.

Reply:
I've implemented the behaviour of falling back to tar in the case of zip files as a separate commit. If the general sentiment is that this fallback behaviour is too complex for the first pass at supporting zip files then I'm happy to remove it.